circuit_passes/src/passes/loop_unroll/loop_env_recorder.rs

```diff
@@ -70,14 +59,24 @@ impl<'a, 'd> EnvRecorder<'a, 'd> {
         EnvRecorder {
             global_data,
             mem,
-            vals_per_iteration: Default::default(),
             current_iter_num: RefCell::new(0),
             safe_to_move: RefCell::new(true),
+            loadstore_to_index_per_iter: Default::default(),
+            vals_at_header_per_iter: Default::default(),
+            env_at_header: RefCell::new(None),
         }
     }
 
-    pub fn get_vals_per_iter(&self) -> Ref<BTreeMap<usize, VariableValues<'a>>> {
-        self.vals_per_iteration.borrow()
+    pub fn take_header_vars_for_iter(&self, iter: &usize) -> BTreeMap<usize, Value> {
+        // 'record_env_at_header' should have been called for each iteration and this
+        //  should only be called once per iteration so the unwrap should never fail.
+        self.vals_at_header_per_iter.borrow_mut().remove(iter).unwrap()
```

Comment:
Change the `unwrap` to an `expect` with the comment. If the invariant doesn't hold for whatever reason it will be quicker to pin point the crash.